# Raw Data NDVI

In [16]:
import pandas as pd

path = "/Users/jincyjose/PycharmProjects/453/old_capstone/Feb22 Final Experiments/training-dataset/raw/corn_ndvi_iowa_state.csv"

df = pd.read_csv(path)
df.head(10)

,NDVI,county_name,date,geoid,year
0,0.071856,Plymouth,2013-01-01,19149,2013
1,0.010591,Woodbury,2013-01-01,19193,2013
2,0.038800,Carroll,2013-01-01,19027,2013
3,0.008795,Dubuque,2013-01-01,19061,2013
4,0.174864,Jefferson,2013-01-01,19101,2013
5,0.122478,Webster,2013-01-01,19187,2013
6,0.134881,Lee,2013-01-01,19111,2013
7,-0.008764,Marshall,2013-01-01,19127,2013
8,0.016668,Worth,2013-01-01,19195,2013
9,0.096305,Cerro Gordo,2013-01-01,19033,2013


# RAW ERA5 WEATHER DATA

In [18]:
import pandas as pd

path = "/Users/jincyjose/PycharmProjects/453/old_capstone/Feb22 Final Experiments/training-dataset/raw/era5_iowa_state.csv"

df = pd.read_csv(path)
df.head(10)

,county_fips,county_name,date,evapotranspiration,month,rainfall,soil_moisture,temperature,year
0,19149,Plymouth,2013-01-01,-0.000111,1,0.000007,0.233207,260.437700,2013
1,19193,Woodbury,2013-01-01,-0.000123,1,0.000011,0.289761,260.778484,2013
2,19027,Carroll,2013-01-01,-0.000098,1,0.000047,0.294660,257.574412,2013
3,19061,Dubuque,2013-01-01,-0.000156,1,0.000003,0.321367,258.797924,2013
4,19101,Jefferson,2013-01-01,-0.000161,1,0.000321,0.330039,261.710081,2013
5,19187,Webster,2013-01-01,-0.000072,1,0.000008,0.309339,255.737033,2013
6,19111,Lee,2013-01-01,-0.000305,1,0.000285,0.330523,263.656307,2013
7,19127,Marshall,2013-01-01,-0.000067,1,0.000033,0.316769,257.313476,2013
8,19195,Worth,2013-01-01,-0.000020,1,0.000008,0.314461,254.251210,2013
9,19033,Cerro Gordo,2013-01-01,-0.000014,1,0.000005,0.317279,254.440343,2013


In [28]:
import pandas as pd

path = "/Users/jincyjose/PycharmProjects/453/old_capstone/Feb22 Final Experiments/training-dataset/raw/IOWA_county_wise_yield.csv"

df = pd.read_csv(path)
df.tail(10)

,Program,Year,Period,Week Ending,Geo Level,State,State ANSI,Ag District,Ag District Code,County,...,Zip Code,Region,watershed_code,Watershed,Commodity,Data Item,Domain,Domain Category,Value,CV (%)
1882,SURVEY,2008,YEAR,NaN,COUNTY,IOWA,19,WEST CENTRAL,40,CARROLL,...,NaN,NaN,0,NaN,CORN,"CORN, GRAIN - YIELD, MEASURED IN BU / ACRE",TOTAL,NOT SPECIFIED,179.0,NaN
1883,SURVEY,2008,YEAR,NaN,COUNTY,IOWA,19,WEST CENTRAL,40,CRAWFORD,...,NaN,NaN,0,NaN,CORN,"CORN, GRAIN - YIELD, MEASURED IN BU / ACRE",TOTAL,NOT SPECIFIED,191.0,NaN
1884,SURVEY,2008,YEAR,NaN,COUNTY,IOWA,19,WEST CENTRAL,40,GREENE,...,NaN,NaN,0,NaN,CORN,"CORN, GRAIN - YIELD, MEASURED IN BU / ACRE",TOTAL,NOT SPECIFIED,159.0,NaN
1885,SURVEY,2008,YEAR,NaN,COUNTY,IOWA,19,WEST CENTRAL,40,GUTHRIE,...,NaN,NaN,0,NaN,CORN,"CORN, GRAIN - YIELD, MEASURED IN BU / ACRE",TOTAL,NOT SPECIFIED,171.0,NaN
1886,SURVEY,2008,YEAR,NaN,COUNTY,IOWA,19,WEST CENTRAL,40,HARRISON,...,NaN,NaN,0,NaN,CORN,"CORN, GRAIN - YIELD, MEASURED IN BU / ACRE",TOTAL,NOT SPECIFIED,176.0,NaN
1887,SURVEY,2008,YEAR,NaN,COUNTY,IOWA,19,WEST CENTRAL,40,IDA,...,NaN,NaN,0,NaN,CORN,"CORN, GRAIN - YIELD, MEASURED IN BU / ACRE",TOTAL,NOT SPECIFIED,196.0,NaN
1888,SURVEY,2008,YEAR,NaN,COUNTY,IOWA,19,WEST CENTRAL,40,MONONA,...,NaN,NaN,0,NaN,CORN,"CORN, GRAIN - YIELD, MEASURED IN BU / ACRE",TOTAL,NOT SPECIFIED,171.0,NaN
1889,SURVEY,2008,YEAR,NaN,COUNTY,IOWA,19,WEST CENTRAL,40,SAC,...,NaN,NaN,0,NaN,CORN,"CORN, GRAIN - YIELD, MEASURED IN BU / ACRE",TOTAL,NOT SPECIFIED,184.0,NaN
1890,SURVEY,2008,YEAR,NaN,COUNTY,IOWA,19,WEST CENTRAL,40,SHELBY,...,NaN,NaN,0,NaN,CORN,"CORN, GRAIN - YIELD, MEASURED IN BU / ACRE",TOTAL,NOT SPECIFIED,190.0,NaN
1891,SURVEY,2008,YEAR,NaN,COUNTY,IOWA,19,WEST CENTRAL,40,WOODBURY,...,NaN,NaN,0,NaN,CORN,"CORN, GRAIN - YIELD, MEASURED IN BU / ACRE",TOTAL,NOT SPECIFIED,185.0,NaN


In [20]:
import pandas as pd

path = "/Users/jincyjose/PycharmProjects/453/old_capstone/Feb22 Final Experiments/training-dataset/raw/iowa_storm_events.csv"

df = pd.read_csv(path)
df.head(10)

,OBJECTID,EPISODE_ID,EVENT_ID,STATE,STATE_FIPS,YEAR,MONTH_NAME,EVENT_TYPE,CZ_TYPE,CZ_FIPS,...,END_AZIMUTH,END_LOCATION,BEGIN_LAT,BEGIN_LON,END_LAT,END_LON,EPISODE_NARRATIVE,EVENT_NARRATIVE,DATA_SOURCE,county
0,1190292,74376,449855,IOWA,19,2013,May,Thunderstorm Wind,C,43,...,N,OSTERDOCK,42.8200,-91.1700,42.8200,-91.1700,"For the second straight day, thunderstorms dev...",An estimated wind gust of 60 mph occurred sout...,CSV,clayton
1,1205829,75276,455094,IOWA,19,2013,July,Thunderstorm Wind,C,75,...,NE,DIKE,42.4800,-92.6100,42.4800,-92.6100,A cold front moved into Iowa during the aftern...,"Four to six 30 foot high trees, 12 inches in d...",CSV,grundy
2,1205857,75276,454848,IOWA,19,2013,July,Thunderstorm Wind,C,169,...,ENE,AMES,42.0400,-93.5700,42.0400,-93.5700,A cold front moved into Iowa during the aftern...,NaN,CSV,story
3,1205988,73897,447284,IOWA,19,2013,May,Thunderstorm Wind,C,113,...,NNW,PARIS,42.2700,-91.6000,42.2700,-91.6000,"A very unstable airmass on May 19, 2013 led to...",A trained spotter measured 76 mph winds from a...,CSV,linn
4,1206024,73914,454665,IOWA,19,2013,May,Thunderstorm Wind,C,139,...,NNW,MUSCATINE,41.4800,-91.0700,41.4800,-91.0700,A cold front pushing into a very moist and uns...,A trained spotter estimated winds of 65 to 75 ...,CSV,muscatine
5,1206025,73914,454666,IOWA,19,2013,May,Thunderstorm Wind,C,45,...,SE,CHARLOTTE,41.9600,-90.4700,41.9600,-90.4700,A cold front pushing into a very moist and uns...,A trained spotter measured 65 mph winds.,CSV,clinton
6,1206078,73751,446211,IOWA,19,2013,June,Thunderstorm Wind,C,77,...,N,STUART,41.5100,-94.3200,41.5100,-94.3200,A low pressure area tracked east out of South ...,Several 6 to 8 inch tree limbs blown down.,CSV,guthrie
7,1206108,75276,455096,IOWA,19,2013,July,Thunderstorm Wind,C,121,...,WNW,BEVINGTON,41.3900,-93.8600,41.3900,-93.8600,A cold front moved into Iowa during the aftern...,Six inch diameter tree branch down across the ...,CSV,madison
8,1206115,73751,446205,IOWA,19,2013,June,Thunderstorm Wind,C,27,...,NW,CARROLL,42.0800,-94.9000,42.0800,-94.9000,A low pressure area tracked east out of South ...,NaN,CSV,carroll
9,1206118,74058,448049,IOWA,19,2013,June,Thunderstorm Wind,C,153,...,SSE,POLK CITY,41.7663,-93.7174,41.7663,-93.7174,An unstable airmass was in place over Iowa wit...,Off duty NWS employee reports 4 inch tree limb...,CSV,polk


# Prepare Features


In [10]:
!python build_new_features.py --mode historical

/Users/jincyjose/PycharmProjects/453/old_capstone/Feb22 Final Experiments/src/features/cleaning.py:115: FutureWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  .apply(_apply_sg)
Processing cutoff: jun01
Final features saved to: /Users/jincyjose/PycharmProjects/453/old_capstone/Feb22 Final Experiments/training-dataset/features_frozen/features_jun01.csv
Processing cutoff: jul01
Final features saved to: /Users/jincyjose/PycharmProjects/453/old_capstone/Feb22 Final Experiments/training-dataset/features_frozen/features_jul01.csv
Processing cutoff: aug01
Final features saved to: /Users/jincyjose/PycharmProjects/453/old_capstone/Feb22 Final Experiments/training-dataset/features_frozen/features_aug01.csv


# April-June Cutoff Features

In [12]:
import pandas as pd

path = "/Users/jincyjose/PycharmProjects/453/old_capstone/Feb22 Final Experiments/training-dataset/features_frozen/features_jun01.csv"

df = pd.read_csv(path)

# ------------------------------------------------
# Basic preview
# ------------------------------------------------
print("First 5 rows:")
display(df.head())

# ------------------------------------------------
# Dataset shape
# ------------------------------------------------
print("Dataset shape (rows, columns):")
print(df.shape)

# ------------------------------------------------
# Column names
# ------------------------------------------------
print("Columns:")
print(list(df.columns))

# ------------------------------------------------
# Data types
# ------------------------------------------------
print("Data types:")
print(df.dtypes)

# ------------------------------------------------
# Missing values
# ------------------------------------------------
print("Missing values per column:")
print(df.isna().sum())

# ------------------------------------------------
# Statistical summary
# ------------------------------------------------
print("Statistical summary:")
display(df.describe())

# ------------------------------------------------
# Unique counties and years
# ------------------------------------------------
print("Unique counties:", df["county"].nunique())
print("Year range:", df["year"].min(), "to", df["year"].max())

First 5 rows:


,county,year,rolling_3yr_mean,yield_bu_acre,lag1_yield,ndvi_peak,ndvi_slope,heat_days_gt32,temp_anomaly,net_moisture_stress,wind_severe_days_58_cutoff
0,adair,2013,132.100000,137.8,104.4,0.490139,0.004365,0,-7.904902,0.573223,0.0
1,adair,2014,131.666667,169.3,137.8,0.548745,0.005757,0,-6.417082,0.222444,0.0
2,adair,2015,137.166667,176.5,169.3,0.549039,0.004720,0,-5.553133,0.373962,0.0
3,adair,2016,161.200000,190.3,176.5,0.553045,0.004054,0,-5.929140,0.251315,0.0
4,adair,2017,178.700000,175.2,190.3,0.553242,0.004564,0,-5.676155,0.155183,1.0



Dataset shape (rows, columns):
(1092, 11)

Columns:
['county', 'year', 'rolling_3yr_mean', 'yield_bu_acre', 'lag1_yield', 'ndvi_peak', 'ndvi_slope', 'heat_days_gt32', 'temp_anomaly', 'net_moisture_stress', 'wind_severe_days_58_cutoff']

Data types:
county                         object
year                            int64
rolling_3yr_mean              float64
yield_bu_acre                 float64
lag1_yield                    float64
ndvi_peak                     float64
ndvi_slope                    float64
heat_days_gt32                  int64
temp_anomaly                  float64
net_moisture_stress           float64
wind_severe_days_58_cutoff    float64
dtype: object

Missing values per column:
county                        0
year                          0
rolling_3yr_mean              0
yield_bu_acre                 0
lag1_yield                    0
ndvi_peak                     0
ndvi_slope                    0
heat_days_gt32                0
temp_anomaly                  0
ne

,year,rolling_3yr_mean,yield_bu_acre,lag1_yield,ndvi_peak,ndvi_slope,heat_days_gt32,temp_anomaly,net_moisture_stress,wind_severe_days_58_cutoff
count,1092.000000,1092.000000,1092.000000,1092.000000,1092.000000,1092.000000,1092.0,1092.000000,1092.000000,1092.000000
mean,2018.289377,180.840476,190.429212,183.775458,0.507519,0.004819,0.0,-6.210478,0.223320,0.246337
std,3.418095,22.519623,21.477838,26.226615,0.053654,0.000872,0.0,0.920650,0.304166,0.540509
min,2013.000000,100.666667,107.300000,85.300000,0.347892,0.002554,0.0,-9.224463,-1.618889,0.000000
25%,2015.000000,165.958333,176.100000,168.775000,0.473658,0.004213,0.0,-6.896689,0.077523,0.000000
50%,2018.000000,185.183333,194.300000,189.450000,0.501391,0.004740,0.0,-6.248010,0.292531,0.000000
75%,2021.000000,198.141667,206.100000,203.150000,0.544543,0.005317,0.0,-5.486014,0.424146,0.000000
max,2024.000000,223.533333,240.900000,234.700000,0.662273,0.007515,0.0,-4.450509,0.690274,4.000000



Unique counties: 98
Year range: 2013 to 2024


# April-July Cutoff Features

In [13]:
import pandas as pd

path = "/Users/jincyjose/PycharmProjects/453/old_capstone/Feb22 Final Experiments/training-dataset/features_frozen/features_jul01.csv"

df = pd.read_csv(path)

# ------------------------------------------------
# Basic preview
# ------------------------------------------------
print("First 5 rows:")
display(df.head())

# ------------------------------------------------
# Dataset shape
# ------------------------------------------------
print("\nDataset shape (rows, columns):")
print(df.shape)

# ------------------------------------------------
# Column names
# ------------------------------------------------
print("\nColumns:")
print(list(df.columns))

# ------------------------------------------------
# Data types
# ------------------------------------------------
print("\nData types:")
print(df.dtypes)

# ------------------------------------------------
# Missing values
# ------------------------------------------------
print("\nMissing values per column:")
print(df.isna().sum())

# ------------------------------------------------
# Statistical summary
# ------------------------------------------------
print("\nStatistical summary:")
display(df.describe())

# ------------------------------------------------
# Unique counties and years
# ------------------------------------------------
print("\nUnique counties:", df["county"].nunique())
print("Year range:", df["year"].min(), "to", df["year"].max())

First 5 rows:


,county,year,rolling_3yr_mean,yield_bu_acre,lag1_yield,ndvi_peak,ndvi_slope,heat_days_gt32,temp_anomaly,net_moisture_stress,wind_severe_days_58_cutoff
0,adair,2013,132.100000,137.8,104.4,0.689283,0.005206,0,-4.779265,0.369262,0.0
1,adair,2014,131.666667,169.3,137.8,0.762252,0.006337,0,-3.575162,0.276065,0.0
2,adair,2015,137.166667,176.5,169.3,0.754628,0.005538,0,-3.001329,0.302114,0.0
3,adair,2016,161.200000,190.3,176.5,0.751038,0.005099,0,-2.521991,-0.101649,0.0
4,adair,2017,178.700000,175.2,190.3,0.743701,0.005244,0,-2.684525,-0.212502,1.0



Dataset shape (rows, columns):
(1092, 11)

Columns:
['county', 'year', 'rolling_3yr_mean', 'yield_bu_acre', 'lag1_yield', 'ndvi_peak', 'ndvi_slope', 'heat_days_gt32', 'temp_anomaly', 'net_moisture_stress', 'wind_severe_days_58_cutoff']

Data types:
county                         object
year                            int64
rolling_3yr_mean              float64
yield_bu_acre                 float64
lag1_yield                    float64
ndvi_peak                     float64
ndvi_slope                    float64
heat_days_gt32                  int64
temp_anomaly                  float64
net_moisture_stress           float64
wind_severe_days_58_cutoff    float64
dtype: object

Missing values per column:
county                        0
year                          0
rolling_3yr_mean              0
yield_bu_acre                 0
lag1_yield                    0
ndvi_peak                     0
ndvi_slope                    0
heat_days_gt32                0
temp_anomaly                  0
ne

,year,rolling_3yr_mean,yield_bu_acre,lag1_yield,ndvi_peak,ndvi_slope,heat_days_gt32,temp_anomaly,net_moisture_stress,wind_severe_days_58_cutoff
count,1092.000000,1092.000000,1092.000000,1092.000000,1092.000000,1092.000000,1092.000000,1092.000000,1092.000000,1092.000000
mean,2018.289377,180.840476,190.429212,183.775458,0.752301,0.006135,0.369963,-3.096031,0.079663,0.494505
std,3.418095,22.519623,21.477838,26.226615,0.042449,0.000789,0.867234,0.799319,0.328280,0.761690
min,2013.000000,100.666667,107.300000,85.300000,0.576478,0.003573,0.000000,-5.718878,-1.242407,0.000000
25%,2015.000000,165.958333,176.100000,168.775000,0.729312,0.005667,0.000000,-3.662977,-0.117969,0.000000
50%,2018.000000,185.183333,194.300000,189.450000,0.755489,0.006205,0.000000,-3.015827,0.137785,0.000000
75%,2021.000000,198.141667,206.100000,203.150000,0.779389,0.006634,0.000000,-2.495748,0.327929,1.000000
max,2024.000000,223.533333,240.900000,234.700000,0.860659,0.008278,7.000000,-1.582235,0.585034,4.000000



Unique counties: 98
Year range: 2013 to 2024


In [14]:
import pandas as pd

path = "/Users/jincyjose/PycharmProjects/453/old_capstone/Feb22 Final Experiments/training-dataset/features_frozen/features_aug01.csv"

df = pd.read_csv(path)

# ------------------------------------------------
# Basic preview
# ------------------------------------------------
print("First 5 rows:")
display(df.head())

# ------------------------------------------------
# Dataset shape
# ------------------------------------------------
print("\nDataset shape (rows, columns):")
print(df.shape)

# ------------------------------------------------
# Column names
# ------------------------------------------------
print("\nColumns:")
print(list(df.columns))

# ------------------------------------------------
# Data types
# ------------------------------------------------
print("\nData types:")
print(df.dtypes)

# ------------------------------------------------
# Missing values
# ------------------------------------------------
print("\nMissing values per column:")
print(df.isna().sum())

# ------------------------------------------------
# Statistical summary
# ------------------------------------------------
print("\nStatistical summary:")
display(df.describe())

# ------------------------------------------------
# Unique counties and years
# ------------------------------------------------
print("\nUnique counties:", df["county"].nunique())
print("Year range:", df["year"].min(), "to", df["year"].max())

First 5 rows:


,county,year,rolling_3yr_mean,yield_bu_acre,lag1_yield,ndvi_peak,ndvi_slope,heat_days_gt32,temp_anomaly,net_moisture_stress,wind_severe_days_58_cutoff
0,adair,2013,132.100000,137.8,104.4,0.794464,0.005007,0,-2.550895,0.123623,0.0
1,adair,2014,131.666667,169.3,137.8,0.868701,0.005793,1,-2.096269,0.012443,0.0
2,adair,2015,137.166667,176.5,169.3,0.873719,0.005343,0,-1.262367,0.263047,0.0
3,adair,2016,161.200000,190.3,176.5,0.885882,0.005190,2,-0.619592,-0.216507,0.0
4,adair,2017,178.700000,175.2,190.3,0.843611,0.004988,5,-0.370429,-0.432878,1.0



Dataset shape (rows, columns):
(1092, 11)

Columns:
['county', 'year', 'rolling_3yr_mean', 'yield_bu_acre', 'lag1_yield', 'ndvi_peak', 'ndvi_slope', 'heat_days_gt32', 'temp_anomaly', 'net_moisture_stress', 'wind_severe_days_58_cutoff']

Data types:
county                         object
year                            int64
rolling_3yr_mean              float64
yield_bu_acre                 float64
lag1_yield                    float64
ndvi_peak                     float64
ndvi_slope                    float64
heat_days_gt32                  int64
temp_anomaly                  float64
net_moisture_stress           float64
wind_severe_days_58_cutoff    float64
dtype: object

Missing values per column:
county                        0
year                          0
rolling_3yr_mean              0
yield_bu_acre                 0
lag1_yield                    0
ndvi_peak                     0
ndvi_slope                    0
heat_days_gt32                0
temp_anomaly                  0
ne

,year,rolling_3yr_mean,yield_bu_acre,lag1_yield,ndvi_peak,ndvi_slope,heat_days_gt32,temp_anomaly,net_moisture_stress,wind_severe_days_58_cutoff
count,1092.000000,1092.000000,1092.000000,1092.000000,1092.000000,1092.000000,1092.000000,1092.000000,1092.000000,1092.000000
mean,2018.289377,180.840476,190.429212,183.775458,0.886666,0.006037,1.288462,-1.167680,-0.080980,0.661172
std,3.418095,22.519623,21.477838,26.226615,0.030735,0.000723,1.839114,0.693341,0.308866,0.918610
min,2013.000000,100.666667,107.300000,85.300000,0.754673,0.003718,0.000000,-3.156025,-1.309231,0.000000
25%,2015.000000,165.958333,176.100000,168.775000,0.870006,0.005659,0.000000,-1.533550,-0.281886,0.000000
50%,2018.000000,185.183333,194.300000,189.450000,0.892770,0.006206,0.000000,-1.124302,-0.020872,0.000000
75%,2021.000000,198.141667,206.100000,203.150000,0.909089,0.006540,2.000000,-0.583338,0.170115,1.000000
max,2024.000000,223.533333,240.900000,234.700000,0.942078,0.007907,11.000000,-0.031379,0.376199,6.000000



Unique counties: 98
Year range: 2013 to 2024


# Best Selected Model

In [8]:
import numpy as np
import pandas as pd
import mlflow
import mlflow.lightgbm
import lightgbm as lgb
import matplotlib.pyplot as plt
import seaborn as sns

from lightgbm import LGBMRegressor
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score
from mlflow.models.signature import infer_signature

from analysis.shap_utils import run_shap_analysis
from analysis.shap_year_contrast import log_shap_year_contrast


# ============================================================
# MAIN MODEL
# ============================================================

def run_lightgbm_limited_features_storm(
        feature_df: pd.DataFrame,
        cutoff_name: str,
        target_col: str = "yield_bu_acre",
):

    df = feature_df.copy()
    df["county"] = df["county"].astype("category")
    years = sorted(df["year"].unique())

    feature_cols = [
        "county", "rolling_3yr_mean", "ndvi_peak", "ndvi_slope",
        "temp_anomaly", "net_moisture_stress",
        "heat_days_gt32", "wind_severe_days_58_cutoff"
    ]

    prediction_rows = []

    # =====================================================
    # WALK FORWARD
    # =====================================================
    for i in range(1, len(years)):

        train_years = years[:i]
        test_year = years[i]

        train_df = df[df["year"].isin(train_years)]
        test_df = df[df["year"] == test_year]

        if test_df.empty:
            continue

        X_train = train_df[feature_cols]
        y_train = train_df[target_col]

        X_test = test_df[feature_cols]
        y_test = test_df[target_col].values

        model = LGBMRegressor(
            n_estimators=1000,
            learning_rate=0.03,
            num_leaves=15,
            min_child_samples=40,
            reg_lambda=5,
            subsample=0.8,
            colsample_bytree=0.8,
            random_state=42,
            n_jobs=-1
        )

        model.fit(
            X_train,
            y_train,
            eval_set=[(X_test, y_test)],
            eval_metric="mae",
            categorical_feature=["county"],
            callbacks=[
                lgb.early_stopping(50, first_metric_only=True),
                lgb.log_evaluation(0)
            ]
        )

        preds = model.predict(X_test)

        fold_pred_df = pd.DataFrame({
            "year": test_year,
            "county": test_df["county"].values,
            "y_true": y_test,
            "y_pred": preds,
        })

        prediction_rows.append(fold_pred_df)

    prediction_df = (
        pd.concat(prediction_rows, ignore_index=True)
        if prediction_rows else
        pd.DataFrame(columns=["year", "county", "y_true", "y_pred"])
    )

    # =====================================================
    # Compute metrics locally (FOR PLOTS ONLY)
    # =====================================================
    overall_mae = overall_rmse = overall_r2 = overall_mape = None

    if not prediction_df.empty:

        overall_r2 = float(r2_score(
            prediction_df["y_true"], prediction_df["y_pred"]
        ))

        overall_rmse = float(np.sqrt(mean_squared_error(
            prediction_df["y_true"], prediction_df["y_pred"]
        )))

        overall_mae = float(mean_absolute_error(
            prediction_df["y_true"], prediction_df["y_pred"]
        ))

        overall_mape = float(
            np.mean(
                np.abs(
                    (prediction_df["y_true"] - prediction_df["y_pred"]) /
                    np.clip(prediction_df["y_true"], 1e-6, None)
                )
            ) * 100
        )

    # =====================================================
    # FINAL MODEL (RETRAIN)
    # =====================================================
    X_full = df[feature_cols]
    y_full = df[target_col]

    final_model = LGBMRegressor(
        n_estimators=1000,
        learning_rate=0.03,
        num_leaves=15,
        min_child_samples=40,
        reg_lambda=5,
        subsample=0.8,
        colsample_bytree=0.8,
        random_state=42,
        n_jobs=-1
    )

    final_model.fit(X_full, y_full, categorical_feature=["county"])

    signature = infer_signature(X_full, final_model.predict(X_full))
    mlflow.lightgbm.log_model(final_model, "model", signature=signature)

    # =====================================================
    # SHAP
    # =====================================================
    shap_values = run_shap_analysis(
        model=final_model,
        X=X_full,
        model_name="LightGBM-Limited-Storm",
        cutoff_name=cutoff_name,
        return_values=True
    )

    log_shap_year_contrast(
        shap_values=shap_values,
        X=X_full,
        years=df["year"],
        cutoff_name=cutoff_name,
        model_name="LightGBM-Limited-Storm"
    )

    # =====================================================
    # DIAGNOSTIC PLOTS RESTORED
    # =====================================================
    if not prediction_df.empty:

        log_production_summary_figure(
            prediction_df,
            shap_values,
            feature_cols,
            overall_mae,
            overall_mape,
            overall_r2,
            "production_summary_storm.png"
        )

        log_residual_diagnostics(
            prediction_df,
            "residual_diagnostics_storm.png"
        )

    return prediction_df, final_model

# ============================================================
# SUMMARY FIGURE
# ============================================================

def log_production_summary_figure(prediction_df, shap_values,
                                  feature_cols,
                                  overall_mae,
                                  overall_mape,
                                  overall_r2,
                                  artifact_name):

    fig, axes = plt.subplots(1, 3, figsize=(20, 6))

    yearly = (
        prediction_df.groupby("year")[["y_true", "y_pred"]]
        .mean()
        .reset_index()
    )

    axes[0].plot(yearly["year"], yearly["y_true"], marker="o", label="Actual")
    axes[0].plot(yearly["year"], yearly["y_pred"], marker="o", label="Predicted")

    if overall_mae is not None:
        axes[0].set_title(
            f"Actual vs Predicted Yield\n"
            f"MAE: {overall_mae:.2f} bu | "
            f"MAPE: {overall_mape:.2f}% | "
            f"R²: {overall_r2:.2f}"
        )
    else:
        axes[0].set_title("Actual vs Predicted Yield")

    axes[0].legend()
    axes[0].grid(True)

    residuals = prediction_df["y_true"] - prediction_df["y_pred"]
    sns.histplot(residuals, bins=25, kde=True, ax=axes[1])
    axes[1].axvline(0, linestyle="--")
    axes[1].set_title("Residual Distribution")
    axes[1].grid(True)

    shap_abs_mean = np.abs(shap_values).mean(axis=0)
    shap_df = pd.DataFrame({
        "feature": feature_cols,
        "importance": shap_abs_mean
    }).sort_values("importance")

    axes[2].barh(shap_df["feature"], shap_df["importance"])
    axes[2].set_title("SHAP Feature Importance")

    plt.tight_layout()
    mlflow.log_figure(fig, artifact_name)
    plt.close(fig)


# ============================================================
# RESIDUAL DIAGNOSTICS
# ============================================================

def log_residual_diagnostics(prediction_df, artifact_name):

    if prediction_df.empty:
        return

    df = prediction_df.copy()
    df["residual"] = df["y_true"] - df["y_pred"]

    fig, axes = plt.subplots(1, 3, figsize=(22, 6))

    sns.histplot(df["residual"], bins=25, kde=True, ax=axes[0])
    axes[0].axvline(0, linestyle="--")
    axes[0].set_title("Overall Residual Distribution")
    axes[0].grid(True)

    county_order = (
        df.groupby("county")["residual"]
        .mean()
        .sort_values()
        .index
    )

    sns.boxplot(
        data=df,
        x="residual",
        y="county",
        order=county_order,
        ax=axes[1]
    )

    axes[1].axvline(0, linestyle="--")
    axes[1].set_title("Residual Distribution by County")

    pivot = df.pivot_table(
        index="county",
        columns="year",
        values="residual"
    )

    sns.heatmap(
        pivot,
        center=0,
        cmap="coolwarm",
        ax=axes[2]
    )

    axes[2].set_title("County-Year Residual Heatmap")

    plt.tight_layout()
    mlflow.log_figure(fig, artifact_name)
    plt.close(fig)

## Train Script

In [7]:
import pandas as pd
import mlflow
import mlflow.lightgbm
import mlflow.sklearn
from mlflow.data import from_pandas
from pathlib import Path
import numpy as np
import json

from sklearn.metrics import r2_score, mean_squared_error, mean_absolute_error

from config import FORECAST_CUTOFFS
from models.model_registry import MODEL_CONFIGS
from analysis.plots import log_county_yield_forecast
from analysis.log_model_comparison_plots import log_comparison_plots

# ============================================================
# PROJECT ROOT
# ============================================================

def find_project_root(start_path: Path):
    for p in [start_path] + list(start_path.parents):
        if (p / "src").exists():
            return p
    raise RuntimeError("Project root not found")

# Works in notebooks and scripts
PROJECT_ROOT = find_project_root(Path.cwd())

FEATURE_DIR = PROJECT_ROOT / "training-dataset" / "features_frozen"
EXPORT_DIR = PROJECT_ROOT / "exported_models"
EXPORT_DIR.mkdir(parents=True, exist_ok=True)


# ============================================================
# SETTINGS
# ============================================================

CUTOFF_STAGE_MAP = {
    "jun01": "Early Season Forecast",
    "jul01": "Mid Season Forecast",
    "aug01": "Late Season Forecast",
}

TARGET_COUNTIES = ["Boone", "Benton", "Marshall", "plymouth", "wayne", "appanoose"] #For sample plotting

MLRUNS_DIR = PROJECT_ROOT / "mlruns"
mlflow.set_tracking_uri(f"file:{MLRUNS_DIR}")
mlflow.set_experiment("iowa_corn_yield_forecasting_feb28")


def _is_lightgbm_model(model) -> bool:
    try:
        from lightgbm.sklearn import LGBMModel
        return isinstance(model, LGBMModel)
    except Exception:
        return model.__class__.__module__.startswith("lightgbm")


def _sanitize_name(s: str) -> str:
    return (
        str(s)
        .strip()
        .replace(" ", "_")
        .replace("/", "_")
        .replace("\\", "_")
        .replace(":", "_")
    )


# ============================================================
# MAIN TRAINING LOOP
# ============================================================

for cutoff_key in FORECAST_CUTOFFS.keys():

    if cutoff_key not in CUTOFF_STAGE_MAP:
        continue

    forecast_stage = CUTOFF_STAGE_MAP[cutoff_key]
    feature_path = FEATURE_DIR / f"features_{cutoff_key}.csv"

    if not feature_path.exists():
        continue

    feature_df = pd.read_csv(feature_path).sort_values(["year", "county"])

    with mlflow.start_run(run_name=forecast_stage):

        mlflow.log_param("cutoff_key", cutoff_key)

        # ====================================================
        # DATASET LOGGING (NEW)
        # ====================================================
        dataset = from_pandas(
            feature_df,
            source=str(feature_path),
            name=f"features_{cutoff_key}"
        )
        mlflow.log_input(dataset, context="training")

        results_dict = {}

        # ====================================================
        # MODEL LOOP
        # ====================================================
        for model_cfg in MODEL_CONFIGS:

            if not model_cfg.get("enabled", False):
                continue

            model_name = model_cfg["name"]

            with mlflow.start_run(run_name=model_name, nested=True):

                result = model_cfg["func"](feature_df, cutoff_key)

                pred_df = None
                final_model = None

                if isinstance(result, tuple):
                    if len(result) >= 1:
                        pred_df = result[0]
                    if len(result) >= 2:
                        final_model = result[1]
                else:
                    pred_df = result

                if pred_df is None or pred_df.empty:
                    continue

                # ====================================================
                # CENTRALIZED METRICS
                # ====================================================
                overall_r2 = float(r2_score(pred_df["y_true"], pred_df["y_pred"]))
                overall_rmse = float(
                    np.sqrt(mean_squared_error(pred_df["y_true"], pred_df["y_pred"]))
                )
                overall_mae = float(
                    mean_absolute_error(pred_df["y_true"], pred_df["y_pred"])
                )
                overall_mape = float(
                    np.mean(
                        np.abs(
                            (pred_df["y_true"] - pred_df["y_pred"])
                            / np.clip(pred_df["y_true"], 1e-6, None)
                        )
                    ) * 100
                )

                mlflow.log_metric("val_r2", overall_r2)
                mlflow.log_metric("val_rmse", overall_rmse)
                mlflow.log_metric("val_mae", overall_mae)
                mlflow.log_metric("val_mape", overall_mape)

                # County-level plots
                for county in TARGET_COUNTIES:
                    log_county_yield_forecast(pred_df, county, model_name)

                results_dict[model_name] = {
                    "pred_df": pred_df,
                    "rmse": overall_rmse,
                    "r2": overall_r2,
                    "mae": overall_mae,
                    "mape": overall_mape,
                    "final_model": final_model,
                }

        # ====================================================
        # SUMMARY TABLE
        # ====================================================
        summary_rows = []
        for model_name, metrics in results_dict.items():
            summary_rows.append(
                {
                    "model": model_name,
                    "rmse": metrics["rmse"],
                    "r2": metrics["r2"],
                    "mae": metrics["mae"],
                    "mape": metrics["mape"],
                }
            )

        summary_df = pd.DataFrame(summary_rows)

        summary_path = f"model_summary_{cutoff_key}.csv"
        summary_df.to_csv(summary_path, index=False)
        mlflow.log_artifact(summary_path)

        # ====================================================
        # SELECT BEST MODEL
        # ====================================================
        best_model_name = None

        if not summary_df.empty:

            best_row = summary_df.loc[summary_df["rmse"].idxmin()]
            best_model_name = str(best_row["model"])

            mlflow.log_metric("best_rmse", float(best_row["rmse"]))
            mlflow.log_metric("best_r2", float(best_row["r2"]))
            mlflow.log_metric("best_mae", float(best_row["mae"]))
            mlflow.log_metric("best_mape", float(best_row["mape"]))
            mlflow.log_param("best_model", best_model_name)

        # ====================================================
        # EXPORT BEST MODEL
        # ====================================================
        if best_model_name and best_model_name in results_dict:

            best_final_model = results_dict[best_model_name]["final_model"]

            if best_final_model is not None:

                cutoff_export_dir = EXPORT_DIR / _sanitize_name(cutoff_key)
                cutoff_export_dir.mkdir(parents=True, exist_ok=True)

                export_path = cutoff_export_dir / _sanitize_name(best_model_name)

                if _is_lightgbm_model(best_final_model):
                    mlflow.lightgbm.save_model(best_final_model, path=str(export_path))
                else:
                    mlflow.sklearn.save_model(best_final_model, path=str(export_path))

                print(f"[OK] Exported best model for {cutoff_key} to: {export_path}")

                # Feature schema export
                if hasattr(best_final_model, "feature_name_"):
                    expected_features = list(best_final_model.feature_name_)
                elif hasattr(best_final_model, "feature_names_in_"):
                    expected_features = list(best_final_model.feature_names_in_)
                else:
                    raise ValueError(
                        f"Model {best_model_name} does not expose feature names."
                    )

                feature_schema = {
                    "cutoff": cutoff_key,
                    "model_name": best_model_name,
                    "expected_features": expected_features,
                    "categorical_features": ["county"] if "county" in expected_features else [],
                    "target": "yield_bu_acre"
                }

                schema_path = export_path / "feature_schema.json"

                with open(schema_path, "w") as f:
                    json.dump(feature_schema, f, indent=4)

                print(f"[OK] Saved feature schema to: {schema_path}")

        # ====================================================
        # COMPARISON PLOTS
        # ====================================================
        plot_input = {
            model_name: metrics["pred_df"]
            for model_name, metrics in results_dict.items()
            if metrics["pred_df"] is not None and not metrics["pred_df"].empty
        }

        log_comparison_plots(results=plot_input, forecast_stage=forecast_stage)

print("Training complete. Metrics logged and models exported.")
print(f"Export folder: {EXPORT_DIR}")

/Users/jincyjose/PycharmProjects/453/old_capstone/venv/lib/python3.11/site-packages/mlflow/data/dataset_source_registry.py:148: UserWarning: The specified dataset source can be interpreted in multiple ways: LocalArtifactDatasetSource, LocalArtifactDatasetSource. MLflow will assume that this is a LocalArtifactDatasetSource source.
  return _dataset_source_registry.resolve(
/Users/jincyjose/PycharmProjects/453/old_capstone/venv/lib/python3.11/site-packages/mlflow/types/utils.py:452: UserWarning: Hint: Inferred schema contains integer column(s). Integer columns in Python cannot represent missing values. If your input data contains missing values at inference time, it will be encoded as floats and will cause a schema enforcement error. The best way to avoid this problem is to infer the model schema based on a realistic data sample (training dataset) that includes missing values. Alternatively, you can declare integer columns as doubles (float64) whenever these columns may have missing valu

[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 0.000188 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 173
[LightGBM] [Info] Number of data points in the train set: 98, number of used features: 6
[LightGBM] [Info] Start training from score 159.446939
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
Training until validation scores don't improve for 50 rounds
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf


/Users/jincyjose/PycharmProjects/453/old_capstone/venv/lib/python3.11/site-packages/mlflow/types/utils.py:452: UserWarning: Hint: Inferred schema contains integer column(s). Integer columns in Python cannot represent missing values. If your input data contains missing values at inference time, it will be encoded as floats and will cause a schema enforcement error. The best way to avoid this problem is to infer the model schema based on a realistic data sample (training dataset) that includes missing values. Alternatively, you can declare integer columns as doubles (float64) whenever these columns may have missing values. See `Handling Integers With Missing Values <https://www.mlflow.org/docs/latest/models.html#handling-integers-with-missing-values>`_ for more details.
  warnings.warn(
2026/03/14 10:34:09 WARNING mlflow.models.model: `artifact_path` is deprecated. Please use `name` instead.
2026/03/14 10:34:09 WARNING mlflow.lightgbm: Saving the models in the pickle or cloudpickle forma

[OK] Exported best model for jun01 to: /Users/jincyjose/PycharmProjects/453/old_capstone/Feb22 Final Experiments/exported_models/jun01/LightGBM-limited_withstorm
[OK] Saved feature schema to: /Users/jincyjose/PycharmProjects/453/old_capstone/Feb22 Final Experiments/exported_models/jun01/LightGBM-limited_withstorm/feature_schema.json

=== Model Comparison ===
                        model        mae       rmse      mape        r2  \
0  LightGBM-limited_withstorm  14.853712  18.493932  7.834268  0.072180   
1              Lag-1 Baseline  17.750805  22.833439  9.417535 -0.414321   
2                       Ridge  18.217013  23.622603  9.724875 -0.513773   

   n_samples  
0        994  
1        994  
2        994  


/Users/jincyjose/PycharmProjects/453/old_capstone/venv/lib/python3.11/site-packages/mlflow/data/dataset_source_registry.py:148: UserWarning: The specified dataset source can be interpreted in multiple ways: LocalArtifactDatasetSource, LocalArtifactDatasetSource. MLflow will assume that this is a LocalArtifactDatasetSource source.
  return _dataset_source_registry.resolve(
/Users/jincyjose/PycharmProjects/453/old_capstone/venv/lib/python3.11/site-packages/mlflow/types/utils.py:452: UserWarning: Hint: Inferred schema contains integer column(s). Integer columns in Python cannot represent missing values. If your input data contains missing values at inference time, it will be encoded as floats and will cause a schema enforcement error. The best way to avoid this problem is to infer the model schema based on a realistic data sample (training dataset) that includes missing values. Alternatively, you can declare integer columns as doubles (float64) whenever these columns may have missing valu

[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 0.000175 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 176
[LightGBM] [Info] Number of data points in the train set: 98, number of used features: 7
[LightGBM] [Info] Start training from score 159.446939
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
Training until validation scores don't improve for 50 rounds
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf


/Users/jincyjose/PycharmProjects/453/old_capstone/venv/lib/python3.11/site-packages/mlflow/types/utils.py:452: UserWarning: Hint: Inferred schema contains integer column(s). Integer columns in Python cannot represent missing values. If your input data contains missing values at inference time, it will be encoded as floats and will cause a schema enforcement error. The best way to avoid this problem is to infer the model schema based on a realistic data sample (training dataset) that includes missing values. Alternatively, you can declare integer columns as doubles (float64) whenever these columns may have missing values. See `Handling Integers With Missing Values <https://www.mlflow.org/docs/latest/models.html#handling-integers-with-missing-values>`_ for more details.
  warnings.warn(
2026/03/14 10:34:24 WARNING mlflow.models.model: `artifact_path` is deprecated. Please use `name` instead.
2026/03/14 10:34:24 WARNING mlflow.lightgbm: Saving the models in the pickle or cloudpickle forma

[OK] Exported best model for jul01 to: /Users/jincyjose/PycharmProjects/453/old_capstone/Feb22 Final Experiments/exported_models/jul01/LightGBM-limited_withstorm
[OK] Saved feature schema to: /Users/jincyjose/PycharmProjects/453/old_capstone/Feb22 Final Experiments/exported_models/jul01/LightGBM-limited_withstorm/feature_schema.json

=== Model Comparison ===
                        model        mae       rmse      mape        r2  \
0  LightGBM-limited_withstorm  13.608702  17.205990  7.228863  0.196909   
1                       Ridge  15.276822  19.012755  8.177775  0.019392   
2              Lag-1 Baseline  17.750805  22.833439  9.417535 -0.414321   

   n_samples  
0        994  
1        994  
2        994  


/Users/jincyjose/PycharmProjects/453/old_capstone/venv/lib/python3.11/site-packages/mlflow/data/dataset_source_registry.py:148: UserWarning: The specified dataset source can be interpreted in multiple ways: LocalArtifactDatasetSource, LocalArtifactDatasetSource. MLflow will assume that this is a LocalArtifactDatasetSource source.
  return _dataset_source_registry.resolve(
/Users/jincyjose/PycharmProjects/453/old_capstone/venv/lib/python3.11/site-packages/mlflow/types/utils.py:452: UserWarning: Hint: Inferred schema contains integer column(s). Integer columns in Python cannot represent missing values. If your input data contains missing values at inference time, it will be encoded as floats and will cause a schema enforcement error. The best way to avoid this problem is to infer the model schema based on a realistic data sample (training dataset) that includes missing values. Alternatively, you can declare integer columns as doubles (float64) whenever these columns may have missing valu

[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 0.000180 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 177
[LightGBM] [Info] Number of data points in the train set: 98, number of used features: 7
[LightGBM] [Info] Start training from score 159.446939
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
Training until validation scores don't improve for 50 rounds
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf


/Users/jincyjose/PycharmProjects/453/old_capstone/venv/lib/python3.11/site-packages/mlflow/types/utils.py:452: UserWarning: Hint: Inferred schema contains integer column(s). Integer columns in Python cannot represent missing values. If your input data contains missing values at inference time, it will be encoded as floats and will cause a schema enforcement error. The best way to avoid this problem is to infer the model schema based on a realistic data sample (training dataset) that includes missing values. Alternatively, you can declare integer columns as doubles (float64) whenever these columns may have missing values. See `Handling Integers With Missing Values <https://www.mlflow.org/docs/latest/models.html#handling-integers-with-missing-values>`_ for more details.
  warnings.warn(
2026/03/14 10:34:39 WARNING mlflow.models.model: `artifact_path` is deprecated. Please use `name` instead.
2026/03/14 10:34:39 WARNING mlflow.lightgbm: Saving the models in the pickle or cloudpickle forma

[OK] Exported best model for aug01 to: /Users/jincyjose/PycharmProjects/453/old_capstone/Feb22 Final Experiments/exported_models/aug01/LightGBM-limited_withstorm
[OK] Saved feature schema to: /Users/jincyjose/PycharmProjects/453/old_capstone/Feb22 Final Experiments/exported_models/aug01/LightGBM-limited_withstorm/feature_schema.json

=== Model Comparison ===
                        model        mae       rmse      mape        r2  \
0  LightGBM-limited_withstorm  10.241370  13.172637  5.456240  0.529293   
1                       Ridge  11.146693  14.363200  5.945215  0.440361   
2              Lag-1 Baseline  17.750805  22.833439  9.417535 -0.414321   

   n_samples  
0        994  
1        994  
2        994  

Training complete. Metrics logged and models exported.
Export folder: /Users/jincyjose/PycharmProjects/453/old_capstone/Feb22 Final Experiments/exported_models


<Figure size 800x500 with 0 Axes>

<Figure size 800x500 with 0 Axes>

<Figure size 800x500 with 0 Axes>

In [27]:
import pandas as pd

base_path = "/Users/jincyjose/PycharmProjects/453/old_capstone/Feb22 Final Experiments/src"

files = {
    "model_summary_jun01.csv": "Planting-Vegetative Model",
    "model_summary_jul01.csv": "Reproductive Model",
    "model_summary_aug01.csv": "Pre-Harvest Model"
}

dfs = []

for file, stage in files.items():
    df = pd.read_csv(f"{base_path}/{file}")
    df["stage_model"] = stage
    dfs.append(df)

combined_df = pd.concat(dfs, ignore_index=True)

# Filter LightGBM
lightgbm_df = combined_df[combined_df["model"].str.contains("light", case=False)]

# Remove model column
lightgbm_df = lightgbm_df.drop(columns=["model"])

# Reorder columns so stage_model comes first
lightgbm_df = lightgbm_df[["stage_model", "rmse", "r2", "mae", "mape"]]

display(lightgbm_df)

,stage_model,rmse,r2,mae,mape
2,Planting-Vegetative Model,18.493932,0.072180,14.853712,7.834268
5,Reproductive Model,17.205990,0.196909,13.608702,7.228863
8,Pre-Harvest Model,13.172637,0.529293,10.241370,5.456240


### Summary of trained models
The yield forecasting system was designed using a stage specific modeling framework that reflects both the progressive availability of data and the biological development of the crop during the growing season. County level datasets were constructed by integrating historical yield records with satellite derived vegetation indicators and weather based stress variables. Key features include rolling three year yield averages, lagged yield, NDVI peak, NDVI growth slope, temperature anomalies, extreme heat days, net moisture stress, and severe wind events. Separate training datasets were created for three forecast cutoffs corresponding to major crop growth phases: the Planting–Vegetative stage (around June 1), the Reproductive stage (around July 1), and the Pre-Harvest stage (around August 1). Candidate models were trained on each dataset and evaluated using RMSE, MAE, MAPE, and the coefficient of determination (R²). MLflow was used to track experiments, ensuring reproducibility of datasets, model parameters, and evaluation metrics.

From an agronomic perspective, predictive performance improves as the season progresses because crop yield becomes increasingly determined by observable plant development and accumulated environmental stress. During the Planting–Vegetative stage, the crop has only recently emerged and canopy development is still limited. At this stage NDVI signals are relatively weak and yield outcomes remain highly sensitive to future weather conditions such as heat waves or drought. Consequently, the early season model achieves modest predictive performance (RMSE ≈ 18.49, R² ≈ 0.07). By the Reproductive stage, the crop enters critical phases such as tasseling and silking, when kernel formation and pollination occur. Vegetation indices become stronger indicators of plant health, and accumulated temperature and moisture stress begin to reflect actual yield potential, resulting in improved performance (RMSE ≈ 17.21, R² ≈ 0.20). Near the Pre-Harvest stage, canopy development has stabilized and grain filling is largely complete, meaning that NDVI patterns, cumulative weather stress, and historical yield signals provide a much clearer representation of final production. As a result, the Pre-Harvest model achieves substantially higher predictive accuracy (RMSE ≈ 13.17, R² ≈ 0.53). This staged modeling approach aligns with the biological progression of crop growth, enabling yield forecasts to become progressively more reliable as the season advances.